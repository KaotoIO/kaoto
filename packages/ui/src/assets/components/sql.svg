<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 19.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 viewBox="0 0 227.2 300" style="enable-background:new 0 0 227.2 300;" xml:space="preserve">
<style type="text/css">
	.st0{fill:#832EAB;}
	.st1{fill:#C32265;}
	.st2{fill:#00AEEF;}
	.st3{fill:none;}
	.st4{fill:#8DC63F;}
	.st5{fill:#0F9BD7;}
	.st6{fill:#FFFFFF;}
	.st7{fill-rule:evenodd;clip-rule:evenodd;fill:#FFFFFF;}
	.st8{fill:#1DA1F2;}
	.st9{fill:#3D5A98;}
	.st10{fill:#231F20;}
	.st11{fill:#007BB5;}
	.st12{fill:#DC4E41;}
	.st13{fill-rule:evenodd;clip-rule:evenodd;fill:#181616;}
	.st14{fill:#181616;}
	.st15{fill:#F2F2F2;}
	.st16{fill:#E3E3E3;}
	.st17{fill:#D54B3D;}
	.st18{fill:#D72B27;}
	.st19{opacity:0.56;fill:url(#SVGID_1_);}
	.st20{fill:#BABABA;}
	.st21{fill:#92D400;}
	.st22{fill:#00B9E4;}
	.st23{fill:#2B3990;}
	.st24{fill:#3F9C35;}
	.st25{fill:#8FCED8;}
	.st26{fill:#D21F21;}
	.st27{fill:#009444;}
	.st28{fill:#4FE89C;}
	.st29{fill:#D6D6D6;}
	.st30{fill:#F1F1F1;}
	.st31{fill:#8C3123;}
	.st32{fill:#E05243;}
	.st33{fill:#5E1F18;}
	.st34{fill:#F2B0A9;}
	.st35{fill:#CACCCE;}
	.st36{fill:#A1AFDB;}
	.st37{fill:#002087;}
	.st38{fill:#6C4199;}
	.st39{fill:#C7C7C7;}
	.st40{fill:#1976D2;}
	.st41{fill:#1E88E5;}
	.st42{fill:#FAFAFA;}
	.st43{fill:#E0E0E0;}
	.st44{fill:#F6A11F;}
	.st45{fill:#0A6DAE;}
	.st46{fill:#0BA25E;}
	.st47{fill:#089156;}
	.st48{fill:#10985B;}
	.st49{fill:#F9C941;}
	.st50{fill:#E8B835;}
	.st51{fill:#296AD9;}
	.st52{fill:#2A71E9;}
	.st53{fill:#286EE6;}
	.st54{fill:#BCBEC0;}
	.st55{fill:#FFFFFF;stroke:#231F20;stroke-width:10;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:10;}
	.st56{fill:#00AEEF;stroke:#231F20;stroke-width:10;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:10;}
	.st57{fill:#231F20;stroke:#231F20;stroke-width:10;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:10;}
	.st58{fill:none;stroke:#231F20;stroke-width:10;stroke-miterlimit:10;}
	.st59{fill:none;stroke:#231F20;stroke-width:10;stroke-linejoin:round;stroke-miterlimit:10;}
	.st60{fill:none;stroke:#231F20;stroke-width:9.5201;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:10;}
	.st61{fill:none;stroke:#231F20;stroke-width:7.8079;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:10;}
	.st62{fill:#D1D3D4;}
	.st63{fill:url(#SVGID_2_);}
	.st64{fill:url(#SVGID_3_);}
	.st65{fill:url(#SVGID_4_);}
	.st66{fill:url(#SVGID_5_);}
	.st67{fill:url(#SVGID_6_);}
	.st68{fill:url(#SVGID_7_);}
	.st69{fill:#00A79D;}
	.st70{fill:#4E376B;}
	.st71{fill:#004981;}
	.st72{fill:#106599;}
	.st73{fill:#1E78AF;}
	.st74{fill:#4396C9;}
	.st75{fill:#63AAD6;}
	.st76{fill:#B8DDF4;}
	.st77{fill:#1F8893;}
	.st78{fill:none;stroke:#000000;stroke-width:6.3625;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:10;}
	.st79{fill:#FFFFFF;stroke:#000000;stroke-width:5.3826;stroke-linejoin:round;stroke-miterlimit:10;}
	.st80{fill:#42BCBC;stroke:#000000;stroke-width:6.2798;stroke-linejoin:round;stroke-miterlimit:10;}
	.st81{fill:none;stroke:#000000;stroke-width:5.3826;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:10;}
	.st82{fill:none;stroke:#000000;stroke-width:3.1813;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:10;}
</style>
<g id="Layer_1">
	<g>
		<g>
			<path class="st27" d="M217.7,194.5c5.9,3.6,9.5,7.7,9.5,12.7v63.4c0,21.6-67.9,29.5-113.6,29.5C67.9,300,0,292.1,0,270.5v-63.4
				c0-4.9,3.5-9.1,9.5-12.7c-5.9-3.6-9.5-7.7-9.5-12.7v-63.4c0-5,3.6-9.2,9.7-12.8c-6.1-3.6-9.8-7.9-9.8-12.9V29.5
				C-0.1,7.9,67.8,0,113.5,0c45.7,0,113.6,7.9,113.6,29.5v63.4c0,5-3.6,9.2-9.7,12.8c6.1,3.6,9.8,7.9,9.8,12.9v63.4
				C227.2,186.7,223.7,190.9,217.7,194.5z"/>
			<path class="st28" d="M217.6,194.7c5.9,3.6-86.1,27.2-208.3,0c-5.9-3.6-9.5-7.7-9.5-12.7v-63.4c0-5,9.7-12.8,9.7-12.8
				s109.6,20,207.7-0.1c9.7,1.8,9.8,7.9,9.8,12.9V182C227.1,186.9,223.6,191.1,217.6,194.7z"/>
			<path class="st6" d="M9.5,105.5C3.5,101.9,0,97.8,0,92.8V29.5C0,7.9,67.9,0,113.6,0c45.7,0,113.6,7.9,113.6,29.5v63.4
				c0,4.9-3.5,9.1-9.5,12.7C114.4,130.1,3.5,109.1,9.5,105.5z"/>
			<path d="M78.1,222.6c12.1,1,24.7,1.5,37.6,1.5c35.5,0,68.7-4,88.9-10.7c1.6-0.5,3.4,0.3,3.9,2c0.5,1.6-0.3,3.4-2,3.9
				c-20.8,6.9-54.8,11-90.9,11c-13,0-25.9-0.5-38.1-1.6c-1.7-0.1-3-1.7-2.8-3.4C74.9,223.8,76.4,222.5,78.1,222.6z M52.9,220.6
				c-12.3-2-22.9-4.7-30.5-7.6c-1.1-0.4-2.3,0.1-2.7,1.2c-0.4,1.1,0.1,2.3,1.2,2.7c8,3.1,18.6,5.8,31.4,7.9c0.1,0,0.2,0,0.3,0
				c1,0,1.9-0.7,2-1.7C54.8,221.9,54,220.8,52.9,220.6z M217.7,194.5c5.9,3.6,9.5,7.7,9.5,12.7v63.4c0,21.6-67.9,29.5-113.6,29.5
				C67.9,300,0,292.1,0,270.5v-63.4c0-4.9,3.5-9.1,9.5-12.7c-5.9-3.6-9.5-7.7-9.5-12.7v-63.4c0-5,3.6-9.2,9.7-12.8
				c-6.1-3.6-9.8-7.9-9.8-12.9V29.5C-0.1,7.9,67.8,0,113.5,0c45.7,0,113.6,7.9,113.6,29.5v63.4c0,5-3.6,9.2-9.7,12.8
				c6.1,3.6,9.8,7.9,9.8,12.9v63.4C227.2,186.7,223.7,190.9,217.7,194.5z M19.3,101.3c3.7,1.5,8.2,3,13.7,4.4
				c17.7,4.6,44.6,8.3,80.5,8.3c36,0,63-3.7,80.6-8.4c5.4-1.4,10-2.9,13.7-4.4c7.2-3,11-6,11-8.4V29.5c0-7.4-36.2-21.2-105.3-21.2
				S8.2,22.1,8.2,29.5v63.4C8.2,95.2,12,98.2,19.3,101.3z M208.3,198.9c-23.5,8.9-64.1,12.4-94.7,12.4c-30.6,0-71.3-3.5-94.7-12.4
				c-7,3-10.6,5.9-10.6,8.2v63.4c0,7.4,36.2,21.2,105.3,21.2s105.3-13.8,105.3-21.2v-63.4C218.9,204.8,215.3,201.9,208.3,198.9z
				 M218.9,118.5c0-2.4-3.8-5.4-11.1-8.4c-23.5,8.8-63.9,12.3-94.3,12.3c-30.4,0-70.7-3.5-94.2-12.2c-7.2,3-11,6-11,8.4v63.4
				c0,2.3,3.6,5.3,10.6,8.2c3.6,1.5,8.1,3,13.5,4.4c17.7,4.7,44.8,8.5,81.2,8.5s63.5-3.8,81.2-8.5c5.4-1.4,9.9-2.9,13.5-4.4
				c7-3,10.6-5.9,10.6-8.2V118.5z M204.6,124.8c-20.2,6.7-53.4,10.7-88.9,10.7c-12.9,0-25.6-0.5-37.6-1.5c-1.7-0.1-3.2,1.1-3.4,2.8
				c-0.1,1.7,1.1,3.2,2.8,3.4c12.2,1,25,1.6,38.1,1.6c36.1,0,70.1-4.1,90.9-11c1.6-0.5,2.5-2.3,2-3.9
				C208,125.1,206.3,124.2,204.6,124.8z M52.9,131.9c-12.3-2-22.9-4.7-30.5-7.6c-1.1-0.4-2.3,0.1-2.7,1.2c-0.4,1.1,0.1,2.3,1.2,2.7
				c8,3.1,18.6,5.8,31.4,7.9c0.1,0,0.2,0,0.3,0c1,0,1.9-0.7,2-1.7C54.8,133.2,54,132.1,52.9,131.9z M77.5,51.2
				c12.2,1,25,1.6,38.1,1.6c36.1,0,70.1-4.1,90.9-11c1.6-0.5,2.5-2.3,2-3.9c-0.5-1.6-2.3-2.5-3.9-2c-20.2,6.7-53.4,10.7-88.9,10.7
				C102.7,46.5,90.1,46,78,45c-1.7-0.1-3.2,1.1-3.4,2.8C74.5,49.5,75.8,51,77.5,51.2z M52.5,47.1c1,0,1.9-0.7,2-1.7
				c0.2-1.1-0.6-2.2-1.7-2.4c-12.3-2-22.9-4.7-30.5-7.6c-1.1-0.4-2.3,0.1-2.7,1.2c-0.4,1.1,0.1,2.3,1.2,2.7c8,3.1,18.6,5.8,31.4,7.9
				C52.2,47.1,52.3,47.1,52.5,47.1z"/>
		</g>
	</g>
</g>
<g id="Layer_2">
</g>
</svg>
