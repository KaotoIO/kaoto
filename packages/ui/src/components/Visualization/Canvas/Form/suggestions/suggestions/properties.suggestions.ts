import { SuggestionProvider } from '@kaoto/forms';

export const propertiesSuggestionProvider: SuggestionProvider = {
  id: 'properties-suggestion-provider',
  appliesTo: (_propName, schema) => schema.type === 'string',
  getSuggestions: async (word, _context) => {
    const normalizedWord = word !== '' ? word : 'foo';

    return [
      // https://camel.apache.org/manual/using-propertyplaceholder.html
      {
        value: `{{${normalizedWord}}}`,
        description: `Use '${normalizedWord}' as a property reference`,
        group: word === '' ? 'Property references' : undefined,
      },
      {
        value: `{{${normalizedWord}:default}}`,
        description: `Use '${normalizedWord}' with a default value as a property reference`,
        group: 'Property references',
      },
      {
        value: `{{?${normalizedWord}}}`,
        description: `Use '${normalizedWord}' as an optional property reference`,
        group: 'Property references',
      },
      {
        value: `{{!${normalizedWord}}}`,
        description: `Use '${normalizedWord}' as a negated property reference`,
        group: 'Property references',
      },
      {
        value: `RAW(${normalizedWord})`,
        description: `Use the RAW value of the '${normalizedWord}' property`,
        group: 'Property references',
      },

      // https://camel.apache.org/manual/using-propertyplaceholder.html#_using_property_placeholder_functions
      {
        value: `{{env:${normalizedWord}}}`,
        description: `Use the '${normalizedWord}' OS environment variable`,
        group: 'Property functions',
      },
      {
        value: `{{sys:${normalizedWord}}}`,
        description: `Use the '${normalizedWord}' Java JVM system property`,
        group: 'Property functions',
      },
      {
        value: `{{service:${normalizedWord.toUpperCase()}}}`,
        description: `Use the ${normalizedWord.toUpperCase()} service, defined using the service naming idiom. Services are expected to be defined as OS ${normalizedWord.toUpperCase()}_SERVICE_HOST and ${normalizedWord.toUpperCase()}_SERVICE_PORT environment variables.`,
        group: 'Property functions',
      },
      {
        value: `{{service.name:${normalizedWord.toUpperCase()}}}`,
        description: `Use the ${normalizedWord.toUpperCase()} service hostname`,
        group: 'Property functions',
      },
      {
        value: `{{service.port:${normalizedWord.toUpperCase()}}}`,
        description: `Use the ${normalizedWord.toUpperCase()} service port`,
        group: 'Property functions',
      },

      // https://camel.apache.org/manual/using-propertyplaceholder.html#_using_kubernetes_property_placeholder_functions
      {
        value: `{{configmap:${normalizedWord}}}`,
        description: `Use the ${normalizedWord} ConfigMap. It should be in the format 'configmap:configmap-name/configmap-key'`,
        group: 'Kubernetes',
      },
      {
        value: `{{configmap:${normalizedWord}}}:defaultValue`,
        description: `Use the ${normalizedWord} ConfigMap with a default value`,
        group: 'Kubernetes',
      },
      {
        value: `{{configmap-binary:${normalizedWord}}}`,
        description: `Use the binary ${normalizedWord} ConfigMap`,
        group: 'Kubernetes',
      },
      {
        value: `{{secret:${normalizedWord}}}`,
        description: `Use the ${normalizedWord} Secret. It should be in the format 'secret:secret-name/secret-key'`,
        group: 'Kubernetes',
      },
      {
        value: `{{secret-binary:${normalizedWord}}}`,
        description: `Use the binary ${normalizedWord} Secret.`,
        group: 'Kubernetes',
      },
    ];
  },
};
