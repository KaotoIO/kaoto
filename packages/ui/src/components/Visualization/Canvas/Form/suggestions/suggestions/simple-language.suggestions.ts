import { Suggestion, SuggestionProvider } from '@kaoto/forms';
import { CamelCatalogService, CatalogKind } from '../../../../../../models';
import { IMetadataApi } from '../../../../../../providers';

const SIMPLE_LANGUAGE_ACTIVATED_FIELDS = [
  '#.message', // This is a special case for the Log EIP
  'simple.expression',
];

// This provider suggests syntax from: https://camel.apache.org/components/next/languages/simple-language.html
export const getSimpleLanguageSuggestionProvider = (metadata?: IMetadataApi['getSuggestions']): SuggestionProvider => {
  return {
    id: 'simple-language-suggestion-provider',
    appliesTo: (propName, schema) => SIMPLE_LANGUAGE_ACTIVATED_FIELDS.includes(propName) && schema.type === 'string',
    getSuggestions: async (word, context) => {
      const normalizedWord = word !== '' ? word : 'name';
      const normalizedWordLowercase = normalizedWord.toLowerCase();

      const simpleLangFunctions = CamelCatalogService.getComponent(CatalogKind.Function, 'simple') ?? {};
      const {
        body,
        ['variable.name']: variableName,
        ['header.name']: headerName,
        ['env.name']: envName,
        ...remainingFunctions
      } = simpleLangFunctions;

      const suggestions: Suggestion[] = [
        { value: `\${variable.${normalizedWord}}`, description: variableName?.description },
        { value: `\${header.${normalizedWord}}`, description: headerName?.description },
      ];

      const bodySuggestion: Suggestion = { value: '${body}', description: body?.description };
      if (word === '') {
        suggestions.unshift(bodySuggestion);
      } else {
        suggestions.push({ ...bodySuggestion, group: 'Simple Language' });
      }

      const environmentVariables = (await metadata?.('env', word, context)) ?? [];
      const environmentVariablesSuggestions = environmentVariables.map((item) => ({
        value: `\${env:${item.value}}`,
        description: `Use the '${item.value}' OS environment variable`,
        group: 'Simple Language: Environment variables',
      }));
      if (environmentVariablesSuggestions.length > 0) {
        suggestions.push(...environmentVariablesSuggestions);
      } else {
        suggestions.push({ value: `\${env.${normalizedWord}}`, description: envName?.description });
      }

      return suggestions.concat(
        Object.entries(remainingFunctions)
          .filter(([name]) => word === '' || name.toLowerCase().startsWith(normalizedWordLowercase))
          .map(([name, func]) => ({
            value: `\${${name}}`,
            description: func.description,
            group: 'Simple Language',
          })),
      );
    },
  };
};
