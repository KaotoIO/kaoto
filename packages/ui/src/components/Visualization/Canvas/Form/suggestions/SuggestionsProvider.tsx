import { useSuggestionRegistry } from '@kaoto/forms';
import { FunctionComponent, PropsWithChildren, useEffect } from 'react';
import { propertiesSuggestionProvider } from './suggestions/properties.suggestions';
import { simpleLanguageSuggestionProvider } from './suggestions/simple-language.suggestions';
import { sqlSyntaxSuggestionProvider } from './suggestions/sql.suggestions';

export const SuggestionRegistrar: FunctionComponent<PropsWithChildren> = ({ children }) => {
  const suggestionRegistryContext = useSuggestionRegistry();

  useEffect(() => {
    suggestionRegistryContext?.registerProvider(propertiesSuggestionProvider);
    suggestionRegistryContext?.registerProvider(simpleLanguageSuggestionProvider);
    suggestionRegistryContext?.registerProvider(sqlSyntaxSuggestionProvider);

    return () => {
      suggestionRegistryContext?.unregisterProvider(propertiesSuggestionProvider.id);
      suggestionRegistryContext?.unregisterProvider(simpleLanguageSuggestionProvider.id);
      suggestionRegistryContext?.unregisterProvider(sqlSyntaxSuggestionProvider.id);
    };
  }, [suggestionRegistryContext]);

  return <>{children}</>;
};
