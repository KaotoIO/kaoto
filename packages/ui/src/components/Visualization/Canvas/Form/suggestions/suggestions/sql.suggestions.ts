// Example of SQL syntax suggestion provider

import { SuggestionProvider } from '@kaoto/forms';
import { simpleLanguageSuggestionProvider } from './simple-language.suggestions';

// This provider suggests syntax from: https://camel.apache.org/components/next/languages/sql.html
export const sqlSyntaxSuggestionProvider: SuggestionProvider = {
  ...simpleLanguageSuggestionProvider,
  id: 'sql-syntax-suggestion-provider',
  appliesTo: (propName, schema) => propName === '#.parameters.query' && schema.type === 'string',
  getSuggestions: async (word, _context) => {
    const suggestionWord = word !== '' ? word : 'foo';

    return [
      {
        value: `:#\${variable.${suggestionWord}}`,
        description: `Reference the '${suggestionWord}' variable`,
      },
      {
        value: `:#\${header.${suggestionWord}}`,
        description: `Reference the '${suggestionWord}' header`,
      },
    ];
  },
};
