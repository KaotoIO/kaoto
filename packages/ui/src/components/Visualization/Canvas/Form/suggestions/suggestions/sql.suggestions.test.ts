import { sqlSyntaxSuggestionProvider } from './sql.suggestions';

describe('Properties Suggestions', () => {
  it.each([
    { propertyName: '#.parameters.query', schema: { type: 'string' }, expected: true },
    { propertyName: 'test', schema: { type: 'string' }, expected: false },
    { propertyName: 'test', schema: { type: 'number' }, expected: false },
    { propertyName: 'test', schema: { type: 'object' }, expected: false },
    { propertyName: 'test', schema: { type: 'array' }, expected: false },
    { propertyName: 'test', schema: { type: 'boolean' }, expected: false },
  ] as const)('should apply to string properties', ({ propertyName, schema, expected }) => {
    const applies = sqlSyntaxSuggestionProvider.appliesTo(propertyName, schema);
    expect(applies).toBe(expected);
  });

  it('should return suggestions', async () => {
    const word = 'test';
    const suggestions = await sqlSyntaxSuggestionProvider.getSuggestions(word, {
      inputValue: 'test example',
      cursorPosition: 0,
      propertyName: 'test-property',
    });

    expect(suggestions).toMatchSnapshot();
  });
});
