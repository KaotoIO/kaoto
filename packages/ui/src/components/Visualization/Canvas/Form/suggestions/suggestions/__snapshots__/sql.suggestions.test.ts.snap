// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Properties Suggestions should return suggestions 1`] = `
[
  {
    "description": "Reference the 'test' variable",
    "value": ":#\${variable.test}",
  },
  {
    "description": "Reference the 'test' header",
    "value": ":#\${header.test}",
  },
]
`;
